fastavro/reader.py

```diff
@@ -134,6 +142,87 @@ def read_boolean(fo, writer_schema=None, reader_schema=None):
     return unpack('B', fo.read(1))[0] != 0
 
 
+def read_timestamp_millis(data, writer_schema=None, reader_schema=None):
+    return datetime.datetime.fromtimestamp(data / 1000).replace(
+        microsecond=data % 1000)
+
+
+def read_timestamp_micros(data, writer_schema=None, reader_schema=None):
+    return datetime.datetime.fromtimestamp(data/1000000).replace(
+        microsecond=data % 1000000)
+
+
+def read_date(data, writer_schema=None, reader_schema=None):
+    return datetime.date.fromordinal(data)
+
+
+def read_bytes_decimal(data, writer_schema=None, reader_schema=None):
+    scale = writer_schema['scale']
+    precision = writer_schema['precision']
+
+    size = len(data)
+
+    if PY3:
+        unscaled_datum = read_decimal_from_fixed(data, precision, scale, size)
+    else:
+        unscaled_datum = read_decimal_from_fixed2(data,
+                                                  precision, scale, size)
+
+    original_prec = getcontext().prec
+    getcontext().prec = precision
+    scaled_datum = Decimal(unscaled_datum).scaleb(-scale)
+    getcontext().prec = original_prec
+    return scaled_datum
+
+
+def read_decimal_from_fixed(datum, precision, scale, size):
+    """
+    Decimal is encoded as fixed. Fixed instances are encoded using the
+    number of bytes declared in the schema.
+    based on https://github.com/apache/avro/pull/82/
+    """
+    unscaled_datum = 0
+    msb = datum[0]
+    leftmost_bit = (msb >> 7) & 1
+    if leftmost_bit == 1:
+        modified_first_byte = (datum[0]) ^ (1 << 7)
+        datum = (modified_first_byte).to_bytes(1, 'big') + datum[1:]
+        for offset in range(size):
+            unscaled_datum <<= 8
+            unscaled_datum += datum[offset]
+        unscaled_datum += pow(-2, (size * 8) - 1)
+    else:
+        for offset in range(size):
+            unscaled_datum <<= 8
+            unscaled_datum += datum[offset]
+
+    return unscaled_datum
+
+
+def read_decimal_from_fixed2(datum, precision, scale, size):
+    """
+    Decimal is encoded as fixed. Fixed instances are encoded using the
+    number of bytes declared in the schema.
+    based on https://github.com/apache/avro/pull/82/
+    """
+    unscaled_datum = 0
+    msb = struct.unpack('!b', datum[0])[0]
+    leftmost_bit = (msb >> 7) & 1
+    if leftmost_bit == 1:
+        modified_first_byte = ord(datum[0]) ^ (1 << 7)
+        datum = chr(modified_first_byte) + datum[1:]
```

Comment:
Is the only line that differs from `read_decimal_from_fixed`? If so maybe we can find something that work on both version.